packages/beacon-state-transition/src/cache/epochContext.ts

```diff
@@ -477,6 +511,60 @@ export class EpochContext {
     return this.proposers[slot % SLOTS_PER_EPOCH];
   }
 
+  getBeaconProposers(): ValidatorIndex[] {
+    return this.proposers;
+  }
+
+  /**
+   * We allow requesting proposal duties 1 epoch in the future as in normal network conditions it's possible to predict
+   * the correct shuffling with high probability. While knowing the proposers in advance is not useful for consensus,
+   * users want to know it to plan manteinance and avoid missing block proposals.
+   *
+   * **How to predict future proposers**
+   *
+   * Proposer duties for epoch N are guaranteed to be known at epoch N. Proposer duties depend exclusively on:
+   * 1. seed (from randao_mix): known 2 epochs ahead
+   * 2. active validator set: known 4 epochs ahead
+   * 3. effective balance: not known ahead
+   *
+   * ```python
+   * def get_beacon_proposer_index(state: BeaconState) -> ValidatorIndex:
+   *   epoch = get_current_epoch(state)
+   *   seed = hash(get_seed(state, epoch, DOMAIN_BEACON_PROPOSER) + uint_to_bytes(state.slot))
+   *   indices = get_active_validator_indices(state, epoch)
+   *   return compute_proposer_index(state, indices, seed)
+   * ```
+   *
+   * **1**: If `MIN_SEED_LOOKAHEAD = 1` the randao_mix used for the seed is from 2 epochs ago. So at epoch N, the seed
+   * is known and unchangable for duties at epoch N+1 and N+2 for proposer duties.
+   *
+   * ```python
+   * def get_seed(state: BeaconState, epoch: Epoch, domain_type: DomainType) -> Bytes32:
+   *   mix = get_randao_mix(state, Epoch(epoch - MIN_SEED_LOOKAHEAD - 1))
+   *   return hash(domain_type + uint_to_bytes(epoch) + mix)
+   * ```
+   *
+   * **2**: The active validator set can be predicted `MAX_SEED_LOOKAHEAD` in advance due to how activations are
+   * processed. We already compute the active validator set for the next epoch to optimize epoch processing, so it's
+   * reused here.
+   *
+   * **3**: Effective balance is not known ahead of time, but it rarely changes. Even if it changes, only a few
+   * balances are sampled to adjust the probability of the next selection (32 per epoch on average). So to invalidate
+   * the prediction the effective of one of those 32 samples should change and change the random_byte inequality.
+   */
+  getBeaconProposersNextEpoch(): ValidatorIndex[] {
```

Comment:
Added a rationale on when and why the proposer prediction stands or is invalidated